src/btree/bt_page.c

```diff
@@ -418,20 +454,178 @@ __wt_page_inmem(WT_SESSION_IMPL *session, WT_REF *ref, const void *image, uint32
+    auxhdr->offset = (uint32_t)WT_PTRDIFF(raw, (uint8_t *)dsk);
+    return (0);
+}
+
 /*
  * __inmem_col_fix --
  *     Build in-memory index for fixed-length column-store leaf pages.
  */
-static void
-__inmem_col_fix(WT_SESSION_IMPL *session, WT_PAGE *page)
+static int
+__inmem_col_fix(WT_SESSION_IMPL *session, WT_PAGE *page, bool *preparedp, size_t *sizep)
 {
     WT_BTREE *btree;
+    WT_CELL_UNPACK_KV unpack;
+    WT_COL_FIX_AUXILIARY_HEADER auxhdr;
     const WT_PAGE_HEADER *dsk;
+    size_t size;
+    uint64_t tmp;
+    uint32_t entry_num, recno_offset, skipped;
+    const uint8_t *p8;
+    bool prepare;
+    void *pv;
 
     btree = S2BT(session);
     dsk = page->dsk;
+    prepare = false;
 
     page->pg_fix_bitf = WT_PAGE_HEADER_BYTE(btree, dsk);
+
+    WT_RET(__wt_col_fix_read_auxheader(session, dsk, &auxhdr));
+
+    switch (auxhdr.version) {
+    case WT_COL_FIX_VERSION_NIL:
+        /* There is no time window data. */
+        page->u.col_fix.fix_tw = NULL;
+        break;
+    case WT_COL_FIX_VERSION_TS:
+        /* The page should be VERSION_NIL if there are no timestamp entries. */
+        WT_ASSERT(session, auxhdr.entries > 0);
+        size = sizeof(WT_COL_FIX_TW) + auxhdr.entries * sizeof(WT_COL_FIX_TW_ENTRY);
+        WT_RET(__wt_calloc(session, 1, size, &pv));
+        *sizep += size;
+        page->u.col_fix.fix_tw = pv;
+        page->pg_fix_numtws = auxhdr.entries;
+
+        recno_offset = 0;
+
+        /* Walk the entries to build the index. */
+        entry_num = 0;
+        WT_CELL_FOREACH_FIX (session, dsk, &auxhdr, unpack) {
+            if (unpack.type == WT_CELL_KEY) {
+                p8 = unpack.data;
+                /* The array is attached to the page, so we don't need to free it on error here. */
+                WT_RET(__wt_vunpack_uint(&p8, unpack.size, &tmp));
+                /* For now at least, check that the entries are in ascending order. */
+                WT_ASSERT(session, tmp < UINT32_MAX);
+                WT_ASSERT(session, (recno_offset == 0 && tmp == 0) || tmp > recno_offset);
+                recno_offset = (uint32_t)tmp;
+            } else if (!WT_TIME_WINDOW_IS_EMPTY(&unpack.tw)) {
+                /* Only index entries that are not already obsolete. */
+                page->pg_fix_tws[entry_num].recno_offset = (uint32_t)recno_offset;
+                page->pg_fix_tws[entry_num].cell_offset = WT_PAGE_DISK_OFFSET(page, unpack.cell);
+                if (unpack.tw.prepare)
+                    prepare = true;
+                entry_num++;
+            }
+        }
+        WT_CELL_FOREACH_END;
+
+        /* Update the actual number of time windows. */
+        page->pg_fix_numtws = entry_num;
+        skipped = auxhdr.entries - entry_num;
+
+        /*
+         * If we skipped _all_ the time windows, discard the index array. Otherwise, assume
+         * reallocating it isn't worthwhile.
+         */
+        if (skipped == auxhdr.entries) {
```

Comment:
It is, but the next set of tests isn't, so I think it's clearer this way. The compiler will take care of the micro-optimization. Anyway, I thought allocating on the fly was going to be a horrible mess, but it's ok, so I'll do that and that eliminates this check. (Done)

The arbitrary criterion is more or less tied to the page size between auxhdr.entries and dsk->u.entries. Tying it harder (e.g. do or don't mark it dirty depending on whether the page is large or small) is possible... but which way should it go? Arguably larger pages should be rewritten less, because they're large and it takes more work, but they're also, at least for normal FLCS pages, also the ones full of time windows where there's the most space to save. (And because we're only marking it, not rewriting it immediately, it's not unlikely that more time windows will expire by the time the page is reconciled and written, especially if there are a lot of them.) So IDK. If you have clearer thoughts on this than I do, I'm happy to change it...

Reply:
I'm good with it the way it is.